### PR TITLE
Configure AsyncHTTPClient Client in http.query() for all scenarios

### LINKED ISSUE #55192: Tornado minion http proxy connection issues
### Description of Issue
There is an ability in Salt to sent minion wide http proxy: 
https://github.com/saltstack/salt/pull/29322
https://github.com/saltstack/salt/pull/43764

Which supports both setting **proxy_host** and **no_proxy** args.
https://docs.saltstack.com/en/latest/ref/configuration/minion.html#proxy-host
https://docs.saltstack.com/en/latest/ref/configuration/minion.html#no-proxy

Setting this args will affect every module and state which is using a **utils.http** since it uses tornado by default and tornado is aware of those proxy options.

The problem is that if you set both **proxy_host** and **no_proxy** option and use it in the same execution run - it **won't work** with a weird bug.

### Setup
Consider setup minion so it will go via http proxy everywhere **except** the 127.0.0.1 - which is a very normal approach.

### Steps to Reproduce Issue
Now try to run this state in this order:
```
test_http_source:
  file.managed:
    - name: /tmp/test.txt
    - source: https://sentry.io
    - skip_verify: True

get_agent_info:
  module.run:
    - consul.agent_self:
      - consul_url: http://127.0.0.1:8500
```
**file.managed** will work just fine and download the contents of the url to the file via proxy. But consul agent which will execute a **consul.get** module which itself will execute the very same **utils.http** lib which will execute the very same **tornado** - will fail will the weird error:
```
'consul.agent_info' failed: initialize() got an unexpected keyword argument 'max_body_size'
```

The problem part of the **utils/http.py** is 
```
        if proxy_host and proxy_port:
            if HAS_CURL_HTTPCLIENT is False:
                ret['error'] = ('proxy_host and proxy_port has been set. This requires pycurl and tornado, '
                                'but the libraries does not seem to be installed')
                log.error(ret['error'])
                return ret

            tornado.httpclient.AsyncHTTPClient.configure('tornado.curl_httpclient.CurlAsyncHTTPClient')
            client_argspec = salt.utils.args.get_function_argspec(
                    tornado.curl_httpclient.CurlAsyncHTTPClient.initialize)
        else:
            client_argspec = salt.utils.args.get_function_argspec(
                    tornado.simple_httpclient.SimpleAsyncHTTPClient.initialize)
        supports_max_body_size = 'max_body_size' in client_argspec.args
```
As you can see we switch from **SimpleAsyncHTTPClient** to **CurlAsyncHTTPClient** only in case of using a proxy. 

So **file.managed** which goes outside via proxy so is using a **CurlAsyncHTTPClient** and it **supports_max_body_size** is set to **False** in that case.

After that, we go to consul which is uses a connection to **127.0.0.1** which is set in **no_proxy** arg, so we fall into else statement and use **SimpleAsyncHTTPClient** and **supports_max_body_size** is **True**.

Yet we get the `TypeError: initialize() got an unexpected keyword argument 'max_body_size'`
So I kinda assume the following:

1. It's not related to consul, it's related to proxy logic vs non-proxy logic
2. Error is triggered by a specific order. You need to run proxy connection first
3. It feels like the non-proxy connection re-use the proxy socket\connection\object but with wrong arguments this time.

### Versions Report
```
Salt Version:
           Salt: 2019.2.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10.3
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Aug  7 2019, 17:28:10)
   python-gnupg: 0.4.5
         PyYAML: 3.11
          PyZMQ: 18.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.1

System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.3.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core
```

### PR DESCRIPTION
**Pull request by Ch3LL**
### What does this PR do?
For the use case of a user calling `salt.utils.http.query` with tornado set as backend and proxy setup on the first call and then the next call you do not use a proxy (for example when calling 127.0.0.1 that is set in no_proxy) you would see this error:
`got an unexpected keyword argument 'max_body_size'`

This is because when using the proxy settings we configure the tornado client to use `curl_httpclient` instead of the default. This PR address this and ensures we also configure the tornado client when not using proxy settings.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/55192

### Previous Behavior
error `got an unexpected keyword argument 'max_body_size`

### New Behavior
No error and we can query http.query with proxy set and then not set in the same run.

### Tests written?
Yes

### Commits signed with GPG?

Yes

ping @s0undt3ch i couldn't get my pre-commit workaround to work. can you run pre-commit here again for me?

### CODE REVIEW

File: requirements/static/linux.in
Comment (Oloremo) [inline requirements/static/linux.in]: You'll have to install `libcurl-devel` or it's distro specific alternative for pycurl to be built.

https://jenkinsci.saltstack.com/job/pr-lint/job/PR-55831/1/console
```
22:05:02      FileNotFoundError: [Errno 2] No such file or directory: 'curl-config': 'curl-config'
```

File: tests/unit/utils/test_http.py
Comment (Oloremo) [inline tests/unit/utils/test_http.py]: I wonder if we can ensure that it will be triggered on some CIs. Right now in a case if there is no pycurl installed on your Jenkins - this test will be never triggered and nobody will even notice that.

**Review by s0undt3ch — DISMISSED** [review]
Requirements have been compiled.

File: requirements/static/linux.in
Comment (Ch3LL) [reply requirements/static/linux.in]: Okay we use the salt-jenkins repo to manage our system dependencies. Ive add this PR here https://github.com/saltstack/salt-jenkins/pull/1630  to add libcurl-dev. I imagine the tests might fail there as the package name could be different across distros. I'll wait and see the results and fix if needed. Once this is merged in we will have to rebuild our golden images, so this might take a while to get all setup correctly.

File: tests/unit/utils/test_http.py
Comment (Ch3LL) [reply tests/unit/utils/test_http.py]: we are ensuring that we install pycurl by adding it as a requirement in `requirements/static/linux.in ` which nox pulls from. The tests failed when attempting to install pycurl, so once libcurl-dev is installed the tests will run on linux distros for sure.

Comment (Oloremo) [reply tests/unit/utils/test_http.py]: oh right, thanks for the clarification!

**Comment by Oloremo** [comment]
@Ch3LL Do you think you could finish this one anytime soon?

**Comment by sagetherage** [comment]
@s0undt3ch can you help with resolving these merge conflicts for Megan, please?

**Comment by s0undt3ch** [comment]
Sure, first thing Monday morning.

**Comment by sagetherage** [comment]
adding this to the Sodium project board with the issue it is tied to in attempt to get this fixed for the release.

**Comment by Oloremo** [comment]
@Ch3LL Mind to rebase this and fix conflicts so it could land in the next release?

**Comment by Ch3LL** [comment]
this is dependent on this getting merged first: https://github.com/saltstack/salt/pull/58192  this will include the images with the correct system dependencies.

**Comment by Ch3LL** [comment]
The debian variants need a couple more dependencies. I added this PR to add those system dependencies: https://github.com/saltstack/salt/pull/55831

**Comment by Ch3LL** [comment]
waiting on this merge https://github.com/saltstack/salt/pull/58706 and then i'll clean up the conflicts and re-run the tests

**Comment by Ch3LL** [comment]
closing in favor of https://github.com/saltstack/salt/pull/58868

**Pull request closed without merging**
